File: RayTracerWindowsUI_VB/RayTracerForm.vb
Imports RayTracerFixtureVB
Imports RayTracerModelVB

Partial Public Class RayTracerForm
    Inherits Form

    Private bitmap As Bitmap
    Private pictureBox As PictureBox
    Const imageWidth As Integer = 600
    Const imageHeight As Integer = 600

    Public Sub New()
        bitmap = New Bitmap(imageWidth, imageHeight)
        pictureBox = New PictureBox()
        pictureBox.Dock = DockStyle.Fill
        pictureBox.SizeMode = PictureBoxSizeMode.StretchImage
        pictureBox.Image = bitmap
        ClientSize = New Drawing.Size(imageWidth, imageHeight + 24)
        Controls.Add(pictureBox)
        Text = "Ray Tracer"
        'Load += AddressOf RayTracerForm_Load
        Show()
    End Sub

    Private Sub RayTracerForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
        Me.Show()
        Dim rayTracer As Renderer = New Renderer(imageWidth, imageHeight, Sub(ByVal x As Integer, ByVal y As Integer, ByVal color As Drawing.Color)
                                                                              bitmap.SetPixel(x, y, color)
                                                                              If x = 0 Then pictureBox.Refresh()
                                                                          End Sub)
        rayTracer.Render(StandardScenes.DefaultScene)
        pictureBox.Invalidate()
    End Sub

    <STAThread>
    Private Shared Sub Main()
        Application.EnableVisualStyles()
        Application.Run(New RayTracerForm())
    End Sub
End Class

File: RayTracerUI/RayTracerForm.cs
using System;
using System.Drawing;
using System.IO;
using System.Windows.Forms;

namespace RayTracer
{
    public partial class RayTracerForm : Form
    {
        Bitmap bitmap;
        PictureBox pictureBox;
        const int imageWidth = 600;
        const int imageHeight = 600;

        public RayTracerForm()
        {
            bitmap = new Bitmap(imageWidth, imageHeight);

            pictureBox = new PictureBox();
            pictureBox.Dock = DockStyle.Fill;
            pictureBox.SizeMode = PictureBoxSizeMode.StretchImage;
            pictureBox.Image = bitmap;

            ClientSize = new System.Drawing.Size(imageWidth, imageHeight + 24);
            Controls.Add(pictureBox);
            Text = "Ray Tracer";
            Load += RayTracerForm_Load;

            Show();
        }

        private void RayTracerForm_Load(object sender, EventArgs e)
        {
            Show();
            Renderer rayTracer = new Renderer(imageWidth, imageHeight);
            Color[] pixels = rayTracer.Render(StandardScenes.DefaultScene);
            Bitmap image = new Bitmap(imageWidth, imageHeight);
            for (int x = 0; x < imageWidth; x++)
            {
                for (int y = 0; y < imageHeight; y++)
                {
                    Color c = pixels[y * imageWidth + x];
                    image.SetPixel(x,y,ToDrawingColor(c));
                }
            }
            pictureBox.Image = image;
            pictureBox.Invalidate();
        }

        private System.Drawing.Color ToDrawingColor(Color c)
        {
            return System.Drawing.Color.FromArgb(c.RedByte(), c.GreenByte(), c.BlueByte());
        }


        [STAThread]
        static void Main()
        {
            Application.EnableVisualStyles();
            Application.Run(new RayTracerForm());
        }
    }
}
